### PR TITLE
Attach addition context to every proposals

### PR DESCRIPTION
**Pull request by overvenus**
It is helpful for special proposals when we want to trace them.

### CODE REVIEW

File: src/raw_node.rs [hunk 1]
```diff
@@ -257,6 +257,11 @@ impl<T: Storage> RawNode<T> {
 
     // Propose proposes data be appended to the raft log.
     pub fn propose(&mut self, data: Vec<u8>, sync_log: bool) -> Result<()> {
+        self.propose_with_context(data, sync_log, 0)
```
Comment (BusyJay) [inline src/raw_node.rs]: I suggest to remove `sync_log` from the `propose` signature, since it's part of the context, which should be handled by users.

**Comment by siddontang** [comment]
can you give an example of how can we use this feature?

File: proto/eraftpb.proto [hunk 2]
```diff
@@ -12,6 +12,7 @@ message Entry {
     uint64 index = 3;
     bytes data = 4;
     bool sync_log = 5;
+    uint64 context = 6;
```
Comment (ngaut) [inline proto/eraftpb.proto]: The context type is uint64, which is very confused.

**Comment by overvenus** [comment]
@siddontang A concrete example can be found at pingcap/tikv#3045, split commands put a flag into the context, so we don't need to deserialize the data field to get types of commands.

@ngaut It is somewhat confusing, it could be `bytes` which compiles to `Vec<u8>`, but I think `u64` is good enough to storage various context, besides `u64` is cheaper than `Vec<u8>`.

**Comment by siddontang** [comment]
@overvenus 

u64 is not enough if we want to support more contexts later.

**Comment by siddontang** [comment]
I prefer using bytes here, and we can use lazy decode in our own logic. 

We can also remove sync-log flag.

**Comment by BusyJay** [comment]
Any updates?

File: src/raw_node.rs [hunk 3]
```diff
@@ -256,29 +256,27 @@ impl<T: Storage> RawNode<T> {
     }
 
     // Propose proposes data be appended to the raft log.
-    pub fn propose(&mut self, data: Vec<u8>, sync_log: bool) -> Result<()> {
+    pub fn propose(&mut self, data: Vec<u8>, context: Vec<u8>) -> Result<()> {
```
Comment (BusyJay) [inline src/raw_node.rs]: `context` should go first.

File: proto/eraftpb.proto [hunk 4]
```diff
@@ -12,6 +12,7 @@ message Entry {
     uint64 index = 3;
     bytes data = 4;
     bool sync_log = 5;
```
Comment (siddontang) [inline proto/eraftpb.proto]: maybe we can add a plan to deprecate sync log later.

**Comment by siddontang** [comment]
Rest LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**